### DISCUSSION TITLE
DFG Fixes: Inconsistencies, Problems, Discussionworthy

### DISCUSSION THREAD

**Discussion by konradweiss** [Fraunhofer-AISEC/cpg#578] [Show and tell]
Specific data flows of the code below are extracted to show how the DFG is build and where problems can be identfied.
# Calls

```java
public class Calls {

    public static void main(String[] args) {
        System.out.println(resolvableCall(1));
        System.out.println(unresolvableCall(2));

        Calls call = new Calls();

        System.out.println(call.resolvableMemberCall(3));
        System.out.println(call.unresolvableMemberCall(4));

    }
    private int resolvableMemberCall(int param){
        return param + 2;
    }
    private static int resolvableCall(int param){
        return param+1;
    }
}
```

## Info
Data flows from args to parameters over expressions are propagated to the return statement and over the `FunctionDeclaration`-Node back to the call.
![DFThroughResolvedFunction](https://user-images.githubusercontent.com/12885000/135816045-dfc407f3-ba27-49bb-9571-ddb10c339fb8.png)
The implicit Constructor is connected to the `ConstructExpression`
![ImplicitConstructorToItsCall](https://user-images.githubusercontent.com/12885000/135816066-f1685120-9ce6-4818-b941-f7b5ffa61832.png)
The implicit return is also connected to the main
![ImplicitReturnToMain](https://user-images.githubusercontent.com/12885000/135816074-bb35f5dc-527c-44d6-b2a5-e1fac495baab.png)
## Problem
IMO there should definitely be a DFG Edge between the `NewExpression` and the contained `ConstructExpression`
![Problem_NoRelationBetweenConstructAndNewE](https://user-images.githubusercontent.com/12885000/135816133-89e99f4f-c679-427f-9960-cc19655a8ccb.png)
## Discussion
For a security Analysis we may want to be on the save side and connect args to the call with a dfg edge if we could not resolve the function target.
![Discussion_NoRelationBetweenArgsAndCallIfResolutionFailed](https://user-images.githubusercontent.com/12885000/135816103-b9dee56d-e582-4ba4-9ddb-2dfd4b8e98e1.png)

# ConditionalOperator

```java
int main() {
  int a = 0;
  int b = 1;

  a = a == b ? b = 2: b = 3;
  a = b;
}
```

## Info
Data used in a condition flows to the root node of the enclosing statement to use its result and know the semantic of what it is used for (the root node represents the context of use).
![ConditionFlowsnoWhereDirectly](https://user-images.githubusercontent.com/12885000/135817298-94cb53d0-718e-46d4-a77d-27cbcb996c40.png)

Data in expressions that use `=` flows over the operator `=`
![ValuesarePropagatedOverAssignmentNodes](https://user-images.githubusercontent.com/12885000/135817311-62e97719-8be5-467a-b957-6b9679f7ee59.png)

## Discussion
Here I see something that could be perceived as inconsistency: The DFs on the left, toward to root node go over the `=`-operator as there it is and expression, while the DFs to the `VariableDeclaration` do not go over the `=` operator. Here we could let the values always flow over the `=`-Operator, and ignore the LHS-Reference. This would also solve a Problem that we see later with the self-operators. However, this is not something that arises from algorithm problem
![Discussion_UseForWriteAndUseInExpressionLookDifferent](https://user-images.githubusercontent.com/12885000/135817325-95683892-848f-44ad-91c2-a06241d2800f.png)

# ControlFlowSensitivity

```java
class ControlFlowSensitive {
    public static void main(String[] args) {
        int a = 0;
        if(a % 2 == 0){
            a = 1;
        }else{
            a = 2;
            System.out.println(a);
        }
        System.out.println(a);
    }
}
```

## Info
Here again we see the condition being evaluated and flowing to the root node of the `IfStatement`
![ConditionIsEvaluatedAndFlowsToIFRoot](https://user-images.githubusercontent.com/12885000/135818709-f50a4488-15bc-4213-ab60-c65cacc17c94.png)

The ControlFlowSensitivity can be seen on the right side where the upper `a` is used in the print outside of the if and has to consider all values written in the branches, and the lower `a` that only has to consider the most recent write in its own branch.
![LeftReferencesOfPrintsShowCFS](https://user-images.githubusercontent.com/12885000/135818719-8ac67dbc-00d2-4ff5-91cd-0484ba4bff9e.png)

# Returns

```java
class Returns {
    public static void main(String[] args) {
        System.out.println(functionThatReturns(1));
    }

    public static int functionThatReturns(int a){
        if(a % 2 == 0){
            return a + 1;
        }else{
            return a;
        }
    }
}
```

## Problem
I identified an issue that apparently only arises when we have multiple returns in a function, only one return has a DFG-Edge to the Function root, and all should have one
![Problem_OnlyOneMaybeFirstReturnsConnectToDeclaration](https://user-images.githubusercontent.com/12885000/135819108-0fe12591-b0e9-4595-8d84-8a1baae660a9.png)

# SelfOperators

```java
public class SelfOperators {

    public static void main(String[] args) {
        int a = 0;
        a++;
        System.out.println(a);

        a += 3;
        System.out.println(a);

    }
}
```

## Problem
Due to some changes the reference of the previous value to the references got lost. In this image we are missing a DFG-Edge from the `Literal` on the left side to the `++` and one from the result of `++` or its reference to the `+=`. This is an issue that I thought we solved but appears to still be there. This is an additional reason to used the operator `=` or `++` and `+=` as data flow start to points-of-use and Declarations.
![Problem_ShouldNotLoseFlowFromLiteral0ToPPandFromPPToPEQ](https://user-images.githubusercontent.com/12885000/135819362-07d23685-5aa8-4b4c-b9da-4bf622d7d8fc.png)

# ImplicitReturn
```
void fun() {
    int a = 2;
}

int main() {
    fun();
    return 0;
}
```

## Discussion
A DF from a function with return type `void`. Nothing is actually returned but we still have a DF. This DF might be acceptable because (1) `void` is a type in some languages and (2) it helps in modelling other function calls.
![Screenshot 2021-10-04 154632](https://user-images.githubusercontent.com/11220054/135863280-8294c038-aef0-476b-b390-237b2ad8aceb.png)

**Comment by fwendland** [comment]
Decisions:

- DFG from `ConstructExpression` to `NewExpression`
- Configuration object for inference engine
  - DFG for unknown functions from parameter into `FunctionDeclaration`
- `this` references are postponed (slightly better implementations for Python and Go; may be adapted)
- Check ConditionalOperator : Is there a dataflow from operator `==` into the conditional operator `?:`?
- DFG for conditional operator
  - Add PropertyEdge support for DFG edges
  - Add implicit dataflow from comparison operator into orange node of coditional expression
- Keep only dataflows from operator to variable (i.e. green to violet to red; no shortcuts directly into variable reference) 
- Returns : May have 3 `return` because one for each branch and an implicit after the `if-else` (which is dead code)?
- Postpone SelfOperator

**Comment by KuechA** [comment]
How do we model currently the data flows for an "unresolvable call"? How is the `base` of such a call involved? In your example above, if we use `String s = call.unresolvableMemberCall(4);`, I believe that there should be a data flow from `call` (and according to the discussion also from `4`) to `s`. I think that this is not the case yet.

Where and how should this be implemented?
